File: sdks/lua/beachcomber/watch_stream.lua
--- WatchStream — a stream of watch events from the beachcomber daemon.
--
-- Returned by Client:watch(). Thin, transport-agnostic wrapper around a
-- backend watch handle (beachcomber.ffi_backend's Watch or
-- beachcomber.subprocess_backend's Watch), both of which expose
-- next_event(timeout_ms)/cancel()/close() returning
-- (event_or_nil, Error_or_nil, outcome_string).

local WatchStream = {}
WatchStream.__index = WatchStream

--- @param handle table  A backend watch handle.
-- @return WatchStream
function WatchStream.new(handle)
  return setmetatable({ _handle = handle, _closed = false }, WatchStream)
end

--- Block for the next event (or return sooner per `timeout_ms`; see the
--- active backend's Watch:next_event for exact semantics — the subprocess
--- transport does not honour a timeout and always blocks for the next line).
-- @param timeout_ms number?  -1 (default) blocks indefinitely, 0 polls, >0 waits that long.
-- @return event table {data, age_ms, stale}, or nil on eof/cancelled, or nil, Error on error
function WatchStream:next_event(timeout_ms)
  if self._closed then return nil end
  local ev, err, outcome = self._handle:next_event(timeout_ms)
  if outcome == "error" then
    return nil, err
  end
  return ev -- nil on timeout/eof/cancelled; the event table on "event"
end

--- Cancel a pending or future next_event() call without closing the handle.
function WatchStream:cancel()
  if self._handle.cancel then self._handle:cancel() end
end

--- Close the watch stream and release the underlying handle.
function WatchStream:close()
  if not self._closed then
    self._closed = true
    self._handle:close()
  end
end

--- Iterator for use in a for-loop: `for event in stream:each() do ... end`.
-- Ends when next_event() returns nil (eof/cancelled/error).
-- @return function iterator
function WatchStream:each()
  return function()
    return self:next_event()
  end
end

return WatchStream

File: sdks/ruby/lib/beachcomber/watch_stream.rb
module Beachcomber
  # Iterates events from a bc_watch_open handle. Create via
  # {Client#watch} rather than directly.
  #
  # bc_watch_next distinguishes five outcomes: event, timeout, eof,
  # cancelled, error. #next_event folds timeout/eof/cancelled into a single
  # nil (matching the previous socket-based API's "nil means the stream is
  # over" contract) and raises the idiomatic exception for error. Pass an
  # explicit +timeout_ms+ to observe timeouts distinctly.
  class WatchStream
    include Enumerable

    def initialize(handle)
      @handle = handle
      @closed = false
    end

    # Yields a WatchEvent per emitted change until the stream ends.
    def each
      return enum_for(:each) unless block_given?

      while (event = next_event)
        yield event
      end
    end

    # Waits for the next event. +timeout_ms+: -1 (default) blocks
    # indefinitely, 0 polls once, >0 waits that long.
    #
    # @return [WatchEvent, nil] nil on end-of-stream (daemon closed the
    #   connection, the watch was cancelled, or the wait elapsed).
    def next_event(timeout_ms = -1)
      json = Beachcomber::FFI.raw_call(:bc_watch_next, @handle, timeout_ms)
      envelope = JSON.parse(json)

      unless envelope['ok']
        err = envelope['error'] || {}
        Beachcomber.raise_for_error(err['kind'] || 'server_error', err['message'] || 'unknown error')
      end

      case envelope['outcome']
      when 'event'
        payload = envelope['data']
        WatchEvent.new(
          data: payload['data'],
          age_ms: (payload['age_ms'] || 0).to_i,
          stale: payload['stale'] == true,
        )
      when 'timeout', 'eof', 'cancelled'
        nil
      else
        raise Beachcomber::ProtocolError, "unknown watch outcome: #{envelope['outcome'].inspect}"
      end
    end

    # Unblocks a pending or future #next_event call. Safe to call from
    # another thread while a call is in flight.
    def cancel
      Beachcomber::FFI.cancel_watch(@handle)
      nil
    end

    def close
      return if @closed

      @closed = true
      Beachcomber::FFI.free_watch(@handle)
    end
  end
end
